File: leet/0720_pre/leet_6.py
#zigzag coversion of a string
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        leng = len(s)
        if leng <= 2 or numRows == 1:
            return s
        # length of a single repeatable pattern
        onelen = 2 * numRows - 2
        #e.g when numRows = 4, rowidx = [0 1 2 3 2 1]
        rowidx = range(numRows) # [0 1 2 3]
        for i in range(numRows, onelen):
            rowidx.append(numRows - 2 - (i % numRows))
        reslst = [''] * numRows #numRow empty lists, each storing one row of chars
        #
        for i in range(leng):
            reslst[rowidx[i % onelen]] += s[i]
        # concat into a single str
        res = '' #empty str
        for i in range(numRows):
            res += reslst[i]
        return res

if __name__ == '__main__':
    solu = Solution()
    print solu.convert("PAYPALISHIRING", 3)
        

File: leet/0719_pre/leet_6.cpp
// zigzag conversion of a string
//convert("PAYPALISHIRING", 3):
//   P   A   H   N
//   A P L S I I G
//   Y   I   R
// result: output row-wisely: "PAHNAPLSIIGYIR"
#include <iostream>
#include <map>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        int len = s.size();
	// this border check is very important
	if(len <= 1 || numRows == 1) return s;
        // length of a single repeated pattern 
        int onelen = 2*numRows - 2;
        // mp stores row indices
        map <int, int> mp;
        for(int i = 0; i < numRows; ++i)
            mp[i] = i;
        for(int i = numRows; i < onelen; ++i)
            mp[i] = numRows - 2 - (i % numRows);
        vector<char> vc; //empty vector of char
        // vector of char vectors, each char vector stores one row of strings
        vector< vector <char> > vvc(numRows, vc); //numRows of empty vc
        for(int i = 0; i < len; ++i)
            vvc[mp[i%onelen]].push_back(s.at(i));
        string res;
        for(int i = 0; i < numRows; ++i)
            res.insert(res.end(), vvc[i].begin(), vvc[i].end());
        return res;
    }
};

int main()
{
	Solution solu;
	cout << solu.convert("aassd", 5) << endl;
	return 0;
}
